scripts/qa_crowbarsetup.sh

```diff
@@ -5038,10 +5039,140 @@ function ping_fips
     done
 }
 
+function get_stack_fips
+{
+    local stack=$1
+    local fip_file=$2
+    local all_fips=$(mktemp)
+    local fip_uuids=$(openstack stack resource list $stack -f value -c physical_resource_id --filter type=OS::Neutron::FloatingIP)
+    openstack floating ip list -f value > $all_fips
+
+    for uuid in $fip_uuids
+        do
+        echo $(awk "\$1 ~ /${uuid}/ { print \$2 }" $all_fips) >> $fip_file
+        done
+
+    rm $all_fips
+}
+
+
+function ping_stack_fips
+{
+    . .openrc
+
+    local stack=$1
+    local fips=$(mktemp)
+
+    get_stack_fips $stack $fips
+
+    for fip in $(cat $fips)
+        do
+        ping -c 1 -w 60 $fip || complain 120 "cannot reach stack $stack floating IP $fip."
+        done
+
+    rm $fips
+}
+
+# Disable all but 3 compute nodes (these are the ones we will deploy the
+# pre-upgrade Heat stack on)
+function oncontroller_reduce_to_3_computes
+{
+    . .openrc
+    echo "Reducing cloud to 3 active compute nodes..."
+    local enable_computes=$(nova service-list | grep nova-compute | awk '{print $2}' | head -n 3)
+    local disable_computes=$(nova service-list | grep nova-compute | awk '{print $2}' | tail -n +4)
+
+    echo "Nodes to enable (first 3):"
+    for node in $enable_computes; do
+        echo "  $node"
+    done
+
+    echo "Nodes to disable: $disable_computes"
+    for node in $disable_computes; do
+        echo "  $node"
+    done
+
+    echo
+    echo "Enabling first 3 nodes..."
+
+    for service in $enable_computes; do
+        nova service-enable $service
+    done
+
+    echo "First 3 nodes enabled."
+
+    echo "Disabling remaining nodes..."
+
+    for service in $disable_computes; do
+        nova service-disable $service
+    done
+
+    echo "Remaining nodes disabled."
+}
+
+# Re-enable the compute nodes we disabled
+function oncontroller_reenable_computes
+{
+    . .openrc
+
+    echo "Re-enabling disabled compute nodes..."
+    for service in $(nova service-list | grep nova-compute | grep disabled | awk '{print $2}'); do
+        nova service-enable $service
+    done
+}
+
+# Use heat_stack_params to provide parameters to heat template
+function oncontroller_testpreupgrade_extended
+{
+    echo "Running extended pre-upgrade test..."
+
+    local tempest_image=$(openstack image list -f value | cut -d ' ' -f 2 | grep '^cirros-.*-tempest$' | head -n 1)
+    local nodes_enabled=$(nova service-list | grep nova-compute | grep enabled | awk '{print $2}')
+    local enabled_count=$(echo $nodes_enabled | wc -w)
+    if [ $enabled_count -lt 3 ]; then
+        complain 11 "testpreupgrade_extended: not enough compute nodes (${enabled_count}). please ensure there are at least 3 enabled compute nodes."
+    fi
```

Comment:
To make it fool proof, you should probably add a check for `$enabled_count -gt 3` as well....

Reply:
Good point. Done.